kernel/src/cpu/irq_state.rs

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: MIT
+//
+// Copyright (c) 2024 SUSE LLC
+//
+// Author: Joerg Roedel <jroedel@suse.de>
+
+use core::arch::asm;
+
+/// Interrupt flag in RFLAGS register
+const EFLAGS_IF: u64 = 1 << 9;
+
+/// Unconditionally disable IRQs
+///
+/// # Safety
+///
+/// Callers need to take care of re-enabling IRQs.
+#[inline(always)]
+pub unsafe fn irqs_disable() {
+    asm!("cli", options(att_syntax));
+}
+
+/// Unconditionally enable IRQs
+///
+/// # Safety
+///
+/// Callers need to make sure it is safe to enable IRQs
+#[inline(always)]
+pub unsafe fn irqs_enable() {
+    asm!("sti", options(att_syntax));
+}
+
+/// Query IRQ state on current CPU
+///
+/// # Returns
+///
+/// `true` when IRQs are enabled, `false` otherwise
+#[inline(always)]
+#[must_use = "Unused irqs_enabled() result - meant to be irq_enable()?"]
+pub fn irqs_enabled() -> bool {
+    // SAFETY: The inline assembly just reads the processors RFLAGS register
+    // and does not change any state.
+    let state: u64 = unsafe {
+        let s: u64;
+        asm!("pushfq",
+             "popq %rax",
+             out("rax") s,
+             options(att_syntax));
```

Comment:
We do not need to force using RAX:

```suggestion
        asm!("pushfq",
             "popq {}",
             out(reg) s,
             options(att_syntax, nomem, preserves_flags));
```

Reply:
Adding `preserves_flags` is only safe if we make sure not to modify `CF`, `PF`, `AF`, `ZF`, `SF`, and `OF` (source: https://doc.rust-lang.org/reference/inline-assembly.html).

Reply:
Updated, thanks.